File: 算法/C++/（3）最大子序列和.cpp
#include <iostream>
#include <vector>

using namespace std;

int getMaxNum(vector<int> data,int left,int right) {

	if (left == right)
	{
		return data[left];
	}

	int midden;
	midden = (left + right) / 2;

	int leftMax = 0;
	int sum = 0;
	for (int i = midden;i >= left; i--)
	{
		sum += data[i];
		if (sum > leftMax)
		{
			leftMax = sum;
		}
	}

	int rightMax = 0;
	sum = 0;
	for (int i = midden+1; i <= right; i++)
	{
		sum += data[i];
		if (sum > rightMax)
		{
			rightMax = sum;
		}
	}

	int leftNum,rightNum;
	leftNum = getMaxNum(data, left, midden);
	rightNum = getMaxNum(data, midden + 1, right);

	int max;
	max = leftNum > rightNum ? leftNum : rightNum;
	if (max > leftMax+rightMax)
	{
		return max;
	}
	else
	{
		return leftMax + rightMax;
	}
}

int main() {

	vector<int> data = {1,2,-3};
	int num = getMaxNum(data, 0, data.size() - 1);
	cout << num << endl;
}

File: 算法/Golang/（3）最大子序列和.go
package main

import "fmt"

func getMaxNum(slice []int,left int,right int)  int{

	if left == right {
		return slice[left]
	}

	midden := (left + right) / 2

	//求包含中点向左序列的最大值
	var leftMax int
	var sum int
	sum = 0
	for i := midden; i >= left; i-- {
		sum += slice[i]
		if sum > leftMax {
			leftMax = sum
		}
	}

	//包含中点向右的最大序列值
	var rightMax int
	sum = 0
	for i := midden + 1; i <= right; i++ {
		sum += slice[i]
		if sum>rightMax {
			rightMax = sum
		}
	}

	leftNum := getMaxNum(slice,left,midden)
	rightNum := getMaxNum(slice,midden+1,right)

	var max int
	if leftNum > rightNum{
		max = leftNum
	}else {
		max = rightNum
	}

	if max > (leftMax+rightMax) {
		return max
	}else {
		return leftMax+rightMax
	}

}

func main()  {

	slice := []int{1,2,-3}
	num := getMaxNum(slice,0,len(slice) - 1)
	fmt.Println(num)
}